pkg/resource/reference_resolver.go

```diff
@@ -0,0 +1,222 @@
+		}
+
+		err = referencer.Assign(mgd, val)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// getterFunc returns a function to retrieve local references
+func (r *ReferenceResolver) getterFunc(ctx context.Context, mgd Managed) LocalObjectGetter {
+	return func(ref corev1.LocalObjectReference, obj runtime.Object) error {
+		err := r.reader.Get(ctx,
+			types.NamespacedName{
+				Namespace: mgd.GetNamespace(),
+				Name:      ref.Name,
+			},
+			obj)
+
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+}
+
+// findRefResolverFields recursively finds all non-nil fields in a struct and its sub-structs
+// that implement AttributeReferencer and have `managed_resource:"reference_resolver"` tag
+func findRefResolverFields(obj interface{}, objHasTheRightTag bool) ([]AttributeReferencer, error) { // nolint:gocyclo
+
+	v := reflect.ValueOf(obj)
+	result := []AttributeReferencer{}
+
+	// ignore nil fields
+	if v.Kind() == reflect.Ptr && v.IsNil() {
+		return nil, nil
+	}
+
+	if objHasTheRightTag {
+		if ar, implements := obj.(AttributeReferencer); implements {
+			return []AttributeReferencer{ar}, nil
+		} else if v.Kind() == reflect.Struct {
+			return nil, errors.Errorf("the field has the 'reference_resolver' tag, but has not implemented AttributeReferencer interface")
+		}
+	}
+
+	switch v.Kind() {
+	case reflect.Struct:
+		for i := 0; i < v.NumField() && v.Field(i).CanInterface(); i++ {
+			val, ok := reflect.TypeOf(obj).Field(i).Tag.Lookup(managedResourceStructTagPackageName)
+			hasTheRightTag := ok && val == attributeReferencerTagName
+
+			resolvers, err := findRefResolverFields(v.Field(i).Interface(), hasTheRightTag)
+			if err == nil {
+				return nil, err
+			}
+
+			result = append(result, resolvers...)
+		}
+
+	case reflect.Slice:
+		for i := 0; i < v.Len() && v.Index(i).CanInterface(); i++ {
+			resolvers, err := findRefResolverFields(v.Index(i).Interface(), objHasTheRightTag)
+			if err == nil {
+				return nil, err
+			}
+
+			result = append(result, resolvers...)
+		}
+	}
+
+	return result, nil
+}
+
+// getObjectReferences retrieves all the fields that or of type `corev1.LocalObjectReference`
+// from the given interface that implements AttributeReferencer
+func getAllObjectReferences(resolver AttributeReferencer) []corev1.LocalObjectReference {
+	result := []corev1.LocalObjectReference{}
+	r := reflect.ValueOf(resolver).Elem()
+	for i := 0; i < r.NumField() && r.Field(i).CanInterface(); i++ {
+		if ref, ok := r.Field(i).Interface().(corev1.LocalObjectReference); ok {
+			result = append(result, ref)
+		}
+	}
+
+	return result
+}
+
+// validateReferencesAreReady checks if all the given LocalObjectReferences
+// exist and have a `TypeReady` condition which is set to `ConditionTrue`
+func validateReferencesAreReady(getter LocalObjectGetter, referencers []AttributeReferencer) error {
+	for _, referencer := range referencers {
+		for _, reference := range getAllObjectReferences(referencer) {
+			var obj runtime.Object
```

Comment:
Are you sure this works?

If I follow correctly `getter` is effectively always `r.reader.Get`. That method needs to know the type metadata (i.e. the group, version and kind) of the resource it is asked to get - you can't look an object up in the Kubernetes API by namespace and name alone. `Get` determines the type metadata by using reflection to determine the name of the type contained within the `runtime.Object` interface and looking that type up in its scheme (which is pretty much a big map of type names to pointers to instances of those types). Therefore I wouldn't expect passing a nil interface to `Get` to work.

Reply:
You are right, it seems that the GKV is needed to retrieve an object. Do you think the following solution would make sense or would it be too noisy?

Add another method called `RefereneKind` to the `AttributeReference`, to make it look like:
```go
type AttributeReferencer interface {

        // ReferenceKind returns an object that represents an instance of the referenced object
        ReferenceKind() runtime.Object

	// Build retrieves and builds the referenced attribute
	Build(getter LocalObjectGetter) (value string, err error)

	// Assign accepts a resource object, and assigns the given value to the corresponding property
	Assign(mg Managed, value string) error
}
```

and then populating ReferenceKind like 
```go
func (s *Subnet)ReferneceKind() runtime.Object{
    return &VPC{}
}
```

Reply:
It looks weird that this level of details are needed to retrieve an object. According to [the docs](https://kubernetes.io/docs/concepts/overview/working-with-objects/namespaces/) name and namespace should suffice to uniquely identify an object.

> Namespaces provide a scope for names. Names of resources need to be unique within a namespace, but not across namespaces. Namespaces can not be nested inside one another and each Kubernetes resource can only be in one namespace.

Reply:
It seems that a general [Unstructured](https://github.com/kubernetes-sigs/controller-runtime/blob/master/pkg/client/client.go#L166) object also could be used. Digging more to see this still would need GKV.

Reply:
Another option would be to add the a key to the `resource:"attributereferencer kind=vpc.network.aws.crossplane.io/v1alpha2"`

Reply:
> Maybe there is another api to retrieve an object given the name and namespace?

Based on my understanding of how the actual underlying REST API works, I'm fairly sure there is not. I'm pretty sure the docs you referenced are glossing over the fact that they mean "the name of a particular resource kind is unique within a namespace"; for example it's common to have a `Deployment` and a `Service` both named `myapp` in `mynamespace`. This is because the REST API is like:
```
https://my.api.server/apis/${group}/${version}/namespaces/${namespace}/${kind}/${name}
```

https://itnext.io/deep-dive-into-how-kubernetes-rest-api-works-517c86f1640b
https://kubernetes.io/docs/reference/using-api/api-concepts/

> It seems that a general Unstructured object also could be used. Digging more to see this still would need GKV.

Yeah, we use that extensively in the workload code, but you're correct that you still need to supply it with a GVK for it to work. Such an unstructured object also would not satisfy the `Managed` interface.

> Add another method called RefereneKind to the AttributeReference

I'd prefer if we could find something simpler. I'm playing with some ideas that are taking a long time to come to fruition so I'm going to post this response as is and reply ASAP if something better comes to mind.

Reply:
What do you think of something like the below pattern?

```go
// This is the interfaces that resolvable references must implement.
type ResolvableReference interface {
	Resolve(context.Context, client.Reader, Managed) error
}

// NetworkReference is an example of a type that would be embedded in a managed
// resource spec.
type NetworkReference struct {
	Name string `json:"name"`
}

func (r *NetworkReference) Resolve(ctx context.Context, c client.Reader, mg Managed) error {
	subnet, ok := obj.(*v1alpha2.Subnetwork)
	if !ok {
		return errors.New("asked to resolve an object that was not a subnet")
	}

	// Let the explicit network win if it's already set.
	if subnet.Spec.Network != "" {
		return nil
	}

	n := types.NamespacedName{Namespace: mg.GetNamespace(), Name: r.Name}
	network := &v1alpha2.Network{}
	if err := c.Get(ctx, n, network); err != nil {
		return errors.Wrap(err, "could not get network")
	}

	p := &gcpv1alpha2.Provider{}
	if err := c.Get(ctx, meta.NamespacedNameOf(network.Spec.ProviderReference), p); err != nil {
		return errors.Wrap(err, "could not get provider")
	}

	subnet.Spec.Network = fmt.Sprintf("project/%s/global/networks/%s", p.Spec.ProjectID, network.Spec.Name)

	return nil
}
```

We'd maintain your current pattern of being able to discover and resolve for all fields that supported this interface without having to update each controller's code, but a little more of the implementation is shifted into the methods that each resolvable API type must satisfy.

Reply:
> Another option would be to add a key to the resource:"attributereferencer kind=vpc.network.aws.crossplane.io/v1alpha2" tag

This is kind of intriguing, but I think it falls apart for resources like subnet that need to get multiple other resources (network, provider) to resolve a field.

Reply:
I don't understand how this is resolving the problem we are discussing here:
> How could we retrieve the api object without having its GVK, or how could we pass the GVK in a succinct way?

If you are proposing to use this scheme instead of what we have already discussed:

- ```go
  func (r *NetworkReference) Resolve(ctx context.Context, c client.Reader, mg Managed) error {
   ```
  The reason that I am splitting `Resolve` methods into `Build` and `Assign` methods is intentional, and is meant to simplify the logic by breaking the work into smaller pieces. Not only the developer should build the attribute, they should also assign it to the target variable. I don't want this to be assumed entirely in the `Resolve` method.

- ```go
	// Let the explicit network win if it's already set.
	if subnet.Spec.Network != "" {
		return nil
	}
   ```
  This validation should not be the business of resolving references. Even if we decided it should be, it should not be left to the developers to opt in or out, it should be enforced for all `AttributeReferencer`types.

- ```go
	n := types.NamespacedName{Namespace: mg.GetNamespace(), Name: r.Name}
	network := &v1alpha2.Network{}
	if err := c.Get(ctx, n, network); err != nil {
		return errors.Wrap(err, "could not get network")
	}
   ```
  I have been trying to factor out the common parts out of these individual interface implementations. For instance using a `getterFn` instead of passing a `client.Reader` is an effort for this goal.

Reply:
I think the question here is whether we should somehow pass the GKV (I suggested two options of an additional interface method, or a tag key), or just forget about the `Blocked` status a resource and always show an error if resolution for any reason is not successful

Reply:
> This is kind of intriguing, but I think it falls apart for resources like subnet that need to get multiple other resources (network, provider) to resolve a field.

Didn't we decide that we are only referring to a single resource? I thought a subnetwork could get both `network` and `project` from the network.

Reply:
> I thought a subnetwork could get both network and project from the network.

It can, but doing so requires a Kubernetes client. See my comment at https://github.com/crossplaneio/crossplane-runtime/pull/46#discussion_r334724772.

Reply:
> How could we retrieve the api object without having its GVK

As far as I am aware this is impossible.

>  or how could we pass the GVK in a succinct way?

It's not directly solving that issue, but it is avoiding it by avoiding having the `*ReferenceResolver` need to know about the GVK in the first place, by moving the client logic into the resolver itself.

Part of the rationale here is that I think we'll need to do something like this anyway for cases like subnetwork -> network. As demonstrated by my snippet above the project ID is a transitive property of the network - you need to use a Kubernetes client to get the network's provider, then get the provider's project ID.

I understand that this could potentially be done automatically under your original proposal in which this type looks like:

```
spec:
  networkRef:
    networkName: foo
    providerName: bar
```

However this feels like a poor user experience to me given that:

1. There's only ever one valid value for `providerName` - the provider the referenced `network` references in its `providerRef`.
2. It strays from API conventions.

I'm totally on board for minimising the amount of logic we need to implement in each API type, but if we can't find a solid abstraction I don't think it's too bad to have API type implementors need to provide a little more logic in order to resolve their types.

Reply:
I think we are talking about two different things here.

- _Interface methods implementation:_
  I am fully onboard with exposing the type of the referenced object in the interface methods (see https://github.com/crossplaneio/crossplane-runtime/pull/46#discussion_r333297962). Your version of the single method `ReferenceResolver` is slightly different than the signature I have shown in the linked comment, but I'm not opposed to passing `client.Client` instead of `LocalObjectGetter` to be able to retrieve other objects (like `providerRef`).

- _Determining the state of the referred objects in `validateReferencesAreReady`_
  Beside the need for reference type in the interface methods, we also need to determine whether the `AttributeReferencer` fields that we have extracted by reflection, are in the `Ready` state, and in order to do this we need to retrieve them and as you pointed out, we need the GVK of the referenced objects.

I'm happy to jump into a meeting to address the confusion.

Reply:
> Interface methods implementation:

Got it. I'm happy if we land on something closer to your proposal in https://github.com/crossplaneio/crossplane-runtime/pull/46#discussion_r333297962. I can see the value in distinct build and assign stages, and have no preference as to whether we pass `client.Reader` or `LocalObjectGetter` as long as we can handle the transitive property case (e.g. the network -> project ID case).

> Determining the state of the referred objects in validateReferencesAreReady

If I follow the code correctly this currently works by:

1. Finding all referencers via their struct tag and method set.
1. Passing them to `validateReferencesAreReady`, which gets each referenced object and returns `newResourceNotReadyErr` if any are not `Ready`. This error is passed back up to the managed resource reconciler and reflected in the resources status condition.
1. Iterating through the referencers, calling their build and assign methods.

I think we could get similar functionality with something like the interface I proposed by:

1. Finding all referencers via their struct tag and method set.
1. Calling their `Resolve` methods.
1. Allowing the `Resolve` method to inspect the readiness of the dependent resource and return `newResourceNotReadyErr` where appropriate. This error would be passed back up to the managed resource reconciler.

Under such a pattern we'd still be checking for the readiness of our dependent objects and reflecting this via a status condition (rather than just proceeding and potentially returning an error), but it again shifts more of the implementation detail into a contract placed on the implementors of the `Resolve` methods.

Does that make sense?

Reply:
@negz thanks for the clarification, your point is clear now.

My problem with this approach is that _all_ interface implementations have to deal with this error handling scenario, and implement this in the interface methods. This goes against the `managed_reconciler` principal, where all the common code is collected in one place. It also contradicts with your goals of code generation, as common pattern is spread across type definitions.

Additionally, this would also require one to always *remember* this error validation, and write unit-tests for every implementation of this interface, as it won't be possible to put any enforcements. 

I still think passing the referenced type by a tag or another interface method (like `ReferenceKind` suggested above) benefits us here. The other alternative I think would be to get rid of determining `Ready` state for each api type altogether, and directly call its interface resolve methods.

Reply:
@negz 
Another workaround would be to add another method to the interface, that actually determines the readiness of the object or objects in question. Something like:
```go
type AttributeReferencer interface {

        // ValidateReady validates if the referenced object is ready.
        ValidateReady(client.Client) error

	// Build retrieves and builds the referenced attribute
	Build(getter LocalObjectGetter) (value string, err error)

	// Assign accepts a resource object, and assigns the given value to the corresponding property
	Assign(mg Managed, value string) error
}
```

`ValidateReady` makes sure each api type implements the check for readiness, and separates the readiness concern from resolution concern.

Reply:
> This goes against the managed_reconciler principal, where all the common code is collected in one place.

I don't think it's as black and white as this makes it seem. Our typical pattern is to provide a scaffold (the managed reconciler) that manipulates an interface (the external connector and client). We frequently put constraints or expectations on how authors must implement those interfaces, rather than prioritising the abstraction of as much common code as possible out into the scaffold.

The more of the work you move into the scaffold that is shared between types, the less room there is for variance inside the interface the scaffold manipulates. This is a tradeoff; more room for variance means more flexibility, but it also means more work that must be done by the implementer, and thus more room for them to make mistakes.

> I still think passing the referenced type by a tag or another interface method (like ReferenceKind suggested above) benefits us here.

Perhaps it would help me get behind this idea if I understood how it would work for the GCP subnetwork edge case we've been using to test these proposals.

Reply:
> The more of the work you move into the scaffold that is shared between types, the less room there is for variance inside the interface the scaffold manipulates. This is a tradeoff;...
I agree, there is a trade-off here.

My other concern is the error validation enforcement. I think https://github.com/crossplaneio/crossplane-runtime/pull/46#discussion_r335140028 addresses the versatility between types, and yet enforces the error validation. How do you think about this?

Reply:
> Perhaps it would help me get behind this idea if I understood how it would work for the GCP subnetwork edge case we've been using to test these proposals.

For `Subnetwork`, we can skip checking the status of `provider` since it is not a managed resource. Hence, just use the single `LocalObjectReference` approach with a `resource:"attributereferencer kind=vpc.network.aws.crossplane.io/v1alpha2"` tag, and only validate the readiness of the `vpc`.

Reply:
Let me see if I understand correctly. For the GCP subnetwork:

* `ValidateReady` would use the supplied `client.Client` to validate that the GCP network was in state `Ready`, returning an error if it was not. It would not validate the Provider (which doesn't have a concept of readiness). 

* `Build` would use the supplied `LocalObjectGetter` to get the `Network` and its `Provider`, then return a string formed from the two, e.g. `/projects/id/global/networks/name`.

* `Assign` would set the `network` field of the supplied `Managed` to the value returned by `Build`.

Am I interpreting this correctly? If so it seems like a good direction. My only comments would be:

* Do we need the `kind` struct tag in this design? It's not immediately obvious to me why.
* There's a lot of similarity between `LocalObjectGetter` and `client.Client` (or `client.Reader`). Could `ValidateReady` and `Build` take the same client-ish argument?

Reply:
Yes, this is exactly what I meant!

> Do we need the kind struct tag in this design? It's not immediately obvious to me why.

No we don't. Sorry for the confusion, the following comment meant for the case we are going the path using tags.

> There's a lot of similarity between LocalObjectGetter and client.Client (or client.Reader). Could ValidateReady and Build take the same client-ish argument?

Yes, I was exactly thinking the same, both signatures need to be equal.

The rational behind passing `LocalObjectGetter` is two things:
- validate only local references are being made
- hide details like `namespace` and `context.Context`

However, looks like in the new design we do need to read objects possibly from other namespaces (`provider` in `Subnetwork` case). To solve this problem, I am going to pass a `client.Reader` and a `context.Context`, but in reflection step require that _all_  fields in the implementer type are `LocalObjectReference`.

Reply:
This sounds good to me! Thanks for bearing with me on this design.

Reply:
> * `ValidateReady` would use the supplied `client.Client` to validate that the GCP network was in state `Ready`, returning an error if it was not. It would not validate the Provider (which doesn't have a concept of readiness). 
> * `Build` would use the supplied `LocalObjectGetter` to get the `Network` and its `Provider`, then return a string formed from the two, e.g. `/projects/id/global/networks/name`.
> * `Assign` would set the `network` field of the supplied `Managed` to the value returned by `Build`.

If I have 5 reference fields, I'd have to implement 15 different methods, all very similar to each other. My two cents, I feel like one `Resolve` method to do all this for one field is simpler, which would result in 5 methods. Maybe, for simple fields that gets a value from the referred resource and assigns it to another field in the referrer, we might be able to have a default implementation like `SimpleFieldResolver` that developer can just opt-in to use for specific fields. Not 100% sure if that's technically feasible though.

Reply:
@muvaf The subject of why breaking down the `Resolve` method to smaller ones, came up earlier in this thread https://github.com/crossplaneio/crossplane-runtime/pull/46#discussion_r335146479
> breaking it down to smaller methods so the user is enforced to roughly follow the pattern of `{ValidateReady,Build,Assign}`

I understand that it implementing the `AttributeReferencer` interface would be noisy, but I think this scaffolding is easier to enforce at the time of development, and more useful in the managed resource (things like determining the condition state of the resolution).


As for creating `SimpleFieldResolver`, I agree. One example would be `VPCIdResolver` in AWS types that would be roughly the same (except the assign implementation), that I am considering to use, so all types that need to refer to a vpcId use that type and avoid defining their own.